screencaptor/src/main/java/com/wealthfront/screencaptor/ScreenCaptor.kt

```diff
@@ -85,199 +84,59 @@ object ScreenCaptor {
-    viewMutators: Set<ViewMutator> = setOf(ScrollbarHider, CursorHider),
-    screenshotDirectory: String = defaultScreenshotDirectory,
-    screenshotFormat: ScreenshotFormat = PNG,
-    screenshotQuality: ScreenshotQuality = BEST
-  ): String {
-    views.forEach { view ->
-      Log.d(SCREENSHOT, "Root view has height(${view.height}) and width(${view.width})")
-      if (view.width == 0 || view.height == 0) {
-        throw IllegalStateException("This view ($view) has no height or width. Is ${view.id} the currently displayed activity?")
-      }
-    }
-
+  ) {
     if (!File(screenshotDirectory).exists()) {
       Log.d(SCREENSHOT, "Creating directory $screenshotDirectory since it does not exist")
       val screenshotDirsCreated = File(screenshotDirectory).mkdirs()
       assert(screenshotDirsCreated)
     }
 
-    val bitmap = createCanvasBitmap(views)
-    val canvas = Canvas(bitmap)
-
-    val viewsToCapture = AtomicInteger(views.size)
     val deviceName = MANUFACTURER.replaceWithUnderscore() + "_" + MODEL.replaceWithUnderscore()
-    val screenshotId = "${screenshotName.toLowerCase(ENGLISH)}_${deviceName}_${SDK_INT}_$screenshotNameSuffix"
-    val screenshotFilePath = "$screenshotDirectory/$screenshotId.${screenshotFormat.extension}"
+    val screenshotId =
+      "${screenshotName.toLowerCase(ENGLISH)}_${deviceName}_${SDK_INT}_$screenshotNameSuffix"
+    val screenshotFile = File("$screenshotDirectory/$screenshotId.${screenshotFormat.extension}")
 
-    mainHandler.post {
+    val initialStateOfViews = hashMapOf<View, ViewTreeState>()
+    getRootViewsFromActivity(activity).forEach { rootView ->
       Log.d(SCREENSHOT, "Modifying view tree for '$screenshotName'")
-      val initialStateOfViews = hashMapOf<View, ViewTreeState>()
-      views.forEach { rootView ->
-        val viewState = modifyViewBeforeScreenshot(
-          rootView,
-          viewDataProcessor,
-          viewMutators,
-          viewModifiers,
-          viewIdsToExclude
-        )
-        initialStateOfViews[rootView] = viewState
+      if (rootView.width == 0 || rootView.height == 0) {
+        throw IllegalStateException("This view ($rootView) has no height or width. Is ${rootView.id} the currently displayed activity?")
       }
 
-      views.forEach { rootView ->
-        rootView.post {
-          Log.d(SCREENSHOT, "Taking screenshot for '$screenshotName'")
-          rootView.draw(canvas)
+      val viewState = modifyViewBeforeScreenshot(
+        rootView,
+        viewDataProcessor,
+        viewMutators,
+        viewModifiers,
+        viewIdsToExclude
+      )
+      initialStateOfViews[rootView] = viewState
+    }
+
+    val screenshotManager = ScreenshotManagerBuilder(activity)
+      .withCustomActionOrder(ScreenshotActionOrder.fallbacksFirst())
+      .build()
 
-          if (viewsToCapture.decrementAndGet() == 0) {
+    mainHandler.post {
+      screenshotManager.makeScreenshot()
+        .observe({ screenshot ->
```

Comment:
since this is async, is there a risk of our test continuing before the screenshot is captured? or will this make use of existing idling resources somehow?

Reply:
this post should be effectively synchronous, since we have Espresso.onIdle() (waits for main looper to be empty) surrounding our screenshot action